homeassistant/components/sunsynkweb/sensor.py

```diff
@@ -0,0 +1,201 @@
+"""Sensors for the sunsynk web api."""
+
+import decimal
+import logging
+
+from homeassistant.components.sensor import (
+    SensorDeviceClass,
+    SensorEntity,
+    SensorStateClass,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import PERCENTAGE, UnitOfEnergy, UnitOfPower
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.update_coordinator import CoordinatorEntity
+
+from . import DOMAIN
+from .coordinator import PlantUpdateCoordinator
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,  # noqa: F821
+) -> None:
+    """Set up sensor devices."""
+    data: PlantUpdateCoordinator = hass.data[DOMAIN][entry.entry_id]
+
+    async_add_entities(
+        [
+            BatteryPowerGauge(data),
+            LoadPowerGauge(data),
+            GridPowerGauge(data),
+            StateOfChargeGauge(data),
+            PVPowerGauge(data),
+            PVEnergyAccumulator(data),
+            LoadAccumulator(data),
+            GridExportAccumulator(data),
+            GridImportAccumulator(data),
+            BatteryChargeAccumulator(data),
+            BatteryDischargeAccumulator(data),
+        ]
+    )
+
+
+class SunSynkApiSensor(CoordinatorEntity, SensorEntity):
+    """Parent class for sunsynk api exposing sensors.
+
+    The sensors expose the sum of the power across all inverters
+    for plants that have more than one inverter.
+    State of charge is normally shared between inverters, so this will take
+    the maximum of the state of charge.
+    """
+
+    _attr_of_interest: str = (
+        ""  # this extra attribute is used as an indirection into the Plant object model
+    )
+
+    def __init__(self, coordinator) -> None:
+        """Initialise the common elements for sunsynk web api sensors."""
+        CoordinatorEntity.__init__(self, coordinator, context=coordinator)
+        self.coordinator: PlantUpdateCoordinator = coordinator
+
+    @callback
+    def _handle_coordinator_update(self) -> None:
+        """Handle updated data from the coordinator."""
+        if self.coordinator.cache is not None:
+            self.native_value = sum(
+                decimal.Decimal(getattr(i, self._attr_of_interest))
+                for i in self.coordinator.cache.plants
+            )
+        else:
+            _LOGGER.debug("Not updating sensor as coordinator cache is empty. ")
+        self.async_write_ha_state()
+
+
+class BatteryPowerGauge(SunSynkApiSensor):
```

Comment:
please use `SensorEntityDescriptions` instead, this will make your code a lot cleaner

Reply:
updated